### DISCUSSION TITLE
ESP32 uasyncio.open_connection latency

### DISCUSSION THREAD

**Discussion by rarefeen** [micropython/micropython#11240] [ESP32]
I'm new to the uasyncio tcp communication, I will try to make it as short as possible. 
I have 3 ESP32, one acts as WiFi router/access point. Two are connected to the router or acts as station.
so I have a server in one of the ESP32. [for example, 192.168.0.3 is server]
whenever I sent data from the other device to the server, I create a asyncio.open_connection connection and send data.

it seems it takes about 1 seconds to make the connection, Is it normal? Since I need to communicate very often to multiple server, making it connect every-time seems quite time consuming which is not feasible in my case.

Is there any workaround or any other idea to do it? I'm trying to communicate between multiple ESP32, some acts as server and client at the same time too. and communication is quite frequent, communication more like star topology.

for simple test, i did the following:
server:
async def send_server(msg, gateway, port):
    print("before conn",utime.ticks_ms()/1000)
    reader, writer = await asyncio.open_connection(str(gateway), port)
    print("after conn",utime.ticks_ms()/1000)
    l = "{:0>4}".format(str(len(str(msg))))
    writer.write("{0}{1}".format(l, msg))
    await writer.drain()
    print("sending",utime.ticks_ms()/1000)
    writer.close()
    await writer.wait_closed()
    print("closed", utime.ticks_ms())

client:
async def server_callback(reader, writer):
    data = await reader.read(4)
    data = await reader.read(int(data.decode()))
    parsed = ujson.loads(data.decode())
    print("data is  ", data.decode())
    writer.close()
    await writer.wait_closed()
    
    
and the output in from the server is [for example]:
before conn 69.283
after conn 70.364
sending 71.157
closed 71161

**Comment by karfas** [comment]
I have no idea why the `connect()` should last a second. Maybe some DNS lookup is involved here ? Try using IP addresses, not names.

Ideas to increase performance:
- for a small number of devices, there is no need to disconnect/close the connection. You might need additional logic to reconnect to the devices that lost connection on another reason.
- Your WLAN client might be configured to use power save options (e.g. PS_MAX_MODEM). This increases latency.
- TCP/IP is not designed for fast/real-time communication. Consider to use another protocol, e.g.
  - UDP
  - UDP broadcasts
  - multicast
  - ESP-NOW

**Reply by rarefeen** [reply]
@karfas thank you for the suggestions, 
I am using IP address directly. since it is a local network, `await asyncio.open_connection(str(gateway), port) `here the gateway variable is 192.168.X.X.
isn't ESP-NOW uses socket? I thought using socket library is a bit difficult with uasyncio when using concurrent/await function. maybe my understanding is wrong.

**Comment by davefes** [comment]
ESPNow does not use sockets.

**Comment by davefes** [comment]
> Sorry I am not very familiar with the ESPNow. Can I use it with uasyncio concurrent co-routine?

Came through on my Gmail account ... I guess it will show-up here later.

Yes.  Have a read through this:
https://micropython-glenn20.readthedocs.io/en/latest/library/espnow.html
Whether or not it is a suitable alternative depends on the amount of data you are trying to send/receive.
250byte packets, simple acknowledgment, but FAST.

https://randomnerdtutorials.com/esp-now-esp32-arduino-ide/

> Default datarate 1Mbs

**Reply by rarefeen** [reply]
Thank you for the link, I will try it soon.
